examples/oauth.rs

```diff
@@ -0,0 +1,261 @@
+    match user {
+        Some(u) => format!(
+            "Hey {}! You're logged in!\nYou may now access `/protected`.\nLog out with `/logout`.",
+            u.username
+        ),
+        None => "You're not logged in.\nVisit `/auth/discord` to do so.".to_string(),
+    }
+}
+
+pub(crate) async fn discord_auth(client: Extension<BasicClient>) -> impl IntoResponse {
+    let (auth_url, _csrf_token) = client
+        .authorize_url(CsrfToken::new_random)
+        .add_scope(Scope::new("identify".to_string()))
+        .url();
+
+    // Redirect to Discord's oauth service
+    Redirect(auth_url.into())
+}
+
+// Valid user session required. If there is none, redirect to the auth page
+pub(crate) async fn protected(user: User) -> impl IntoResponse {
+    format!(
+        "Welcome to the protected area :)\nHere's your info:\n{:?}",
+        user
+    )
+}
+
+pub(crate) async fn logout(
+    store: Extension<MemoryStore>,
+    cookies: TypedHeader<headers::Cookie>,
+) -> impl IntoResponse {
+    let cookie = cookies.get(COOKIE_NAME).unwrap();
+    let session = match store.load_session(cookie.to_string()).await.unwrap() {
+        Some(s) => s,
+        // No session active, just redirect
+        None => return Redirect("/".to_string()),
+    };
+
+    store.destroy_session(session).await.unwrap();
+
+    Redirect("/".to_string())
+}
+
+#[derive(Debug, Deserialize)]
+pub struct AuthRequest {
+    pub code: String,
+    pub state: String,
+}
+
+pub async fn login_authorized(
+    query: Query<AuthRequest>,
+    store: Extension<MemoryStore>,
+    oauth_client: Extension<BasicClient>,
+) -> impl IntoResponse {
+    // Get an auth token
+    let token = oauth_client
+        .exchange_code(AuthorizationCode::new(query.code.clone()))
+        .request_async(async_http_client)
+        .await
+        .unwrap();
+
+    // Fetch user data from discord
+    let client = reqwest::Client::new();
+    let user_data: User = client
+        // https://discord.com/developers/docs/resources/user#get-current-user
+        .get("https://discordapp.com/api/users/@me")
+        .bearer_auth(token.access_token().secret())
+        .send()
+        .await
+        .unwrap()
+        .json::<User>()
+        .await
+        .unwrap();
+
+    // Create a new session filled with user data
+    let mut session = Session::new();
+    session.insert("user", &user_data).unwrap();
+
+    // Store session and get corresponding cookie
+    let cookie = store.0.store_session(session).await.unwrap().unwrap();
+
+    let cookie = Cookie::build(COOKIE_NAME, cookie)
+        .same_site(SameSite::Lax)
+        .path("/")
+        .finish();
+
+    // Set cookie
+    let r = http::Response::builder()
+        .header("Location", "/")
+        .header(SET_COOKIE, cookie.to_string())
+        .status(302);
+
+    r.body(Body::empty()).unwrap()
+}
+
+// Utility to save some lines of code
+pub struct Redirect(String);
+impl<'a> IntoResponse for Redirect {
```

Comment:
```suggestion
impl IntoResponse for Redirect {
```

Doesn't seem to be used.